File: 6_datenerhebung/6_1_webscraping/webscraping.py
# Bibliotheken einbinden
import os
import requests
import pandas as pd
from bs4 import BeautifulSoup

# Unterverzeichnis zum Herunterladen festlegen und
# anlegen, wenn es noch nicht existiert
directory = "html"
if not os.path.exists(directory):
    os.makedirs(directory)

# URL und Dateiname festlegen
url = "http://www.fernsehserien.de/serien-a-z/n"
dateiname = directory+"/fernsehserien-g.html"    

# Herunterladen
response = requests.get(url)  
if response.status_code == 200:
    with open(dateiname, 'wb') as datei:
        datei.write(response.content)    

# Datei öffnen und Html mit Beautifulsoup parsen
soup = BeautifulSoup(open(dateiname,encoding="utf-8"),'lxml')

# Das ul-Element mit der ID a-z-liste finden,
# darin alle li-Elemente finden
soup_ul = soup.find('ul',{'id':'a-z-liste'})
soup_li = soup_ul.find_all('li')
 
# Alle li-Elemente abarbeiten,
# die Ergebnisse in der results-Liste
# ablegen
results = []

for li in soup_li:
    # Ein leeres dict anlegen, in dem die Werte 
    # gespeichert werden
    # Name der Serie aus span-Tag auslesen
    # Das dict zur Liste hinzufügen
    item = {}
    item ['name'] = li.find('span').text 
    results.append(item)
  

# Liste mit Dictionaries in DataFrame umwandeln
results = pd.DataFrame(results)
# Diese Liste im Notebook ausgeben 
# (erste und letzte Einträge)

pd.set_option('display.max_rows', 10)
display(results)

# DataFrame als CSV-Datei abpseichern
results.to_csv('results.csv',sep=";",index=False)

File: 6_datenerhebung/6_1_webscraping/webscraping.R
#
# Klassisches Webscraping mit Rvest
#

# Libraries
library(tidyverse)
library(httr)
library(rvest)


# Seite herunterladen und parsen
url <- "http://www.fernsehserien.de/serien-a-z/n"
html <- read_html(url)
html

# Das geparste Dokument kann bei Bedarf abgespeichert werden
write_xml(html, file="serien_n.html")

# Alternative:
# - Webseite  mit GET-Funktion aus dem httr packager herunterladen
# - Inhalt in HTML-Datei abspeichern
# - Datei mit rvest einlesen
response <- GET(url)
cat(content(response, "text"), file="serien_n.html")
html <- read_html("serien_n.html")


# Mithilfe von CSS das ul-Element 
# mit der ID "a-z-liste" und darin alle li-Elemente finden
lis <- html %>% html_nodes("ul#a-z-liste li")
lis


# Leerer Vektor zum Sammeln der Daten
serien <- c()


# Alle li-Elemente abarbeiten
for (li in lis) {
  
  # - Daten aus dem li-Element auslesen
  li_name <- html_node(li,"span") %>% html_text()
  li_jahr <- html_attr(li,"data-jahr")
  li_link <- html_node(li,"a") %>% html_attr("href")
  li_html <- as.character(li)
  
  # - in einem neuen Tibble ablegen
  serie <- tibble('name'=li_name,
                  'jahr'=li_jahr,
                  'link'=li_link,
                  'html'=li_html)
  
  # - Tibble zum Vektor hinzufügen
  serien <-  c(serien,list(serie))
  
}


# Die Liste mit allen Tibbles zu einem 
# einzigen Tibble verbinden
serien <- bind_rows(serien)

# Ergebnis als Excel-Datei abspeichern
write_xlsx(serien,"serien_n.xlsx")
